### PR TITLE
Ports/psoc6 jensb

### PR DESCRIPTION
**Pull request by bargfred**

### CODE REVIEW

**Comment by boramonideep** [comment]
@jaenrig-ifx:  Looks like some functions from the `libs` folder are used in this implementation so we have to clone the submodules inside as well. I guess you need to add `make submodules` to the docker flow script(s) as well, before compiling everything with make.

**Comment by jaenrig-ifx** [comment]
Added `make submodules` to the ci/cd and run codeformatting tool.

File: ports/psoc6/boards/manifest.py [hunk 1]
```diff
@@ -0,0 +1,6 @@
+freeze("$(PORT_DIR)/freeze")
+# include("$(MPY_DIR)/extmod/uasyncio")
```
Comment (jaenrig-ifx) [inline ports/psoc6/boards/manifest.py:2]: Remove all these comments? nothing of this is needed or?

File: ports/psoc6/main.c [hunk 2]
```diff
@@ -86,14 +107,21 @@ int main(int argc, char **argv) {
...
+
+    mp_hal_stdout_tx_str(MICROPY_BANNER_NAME_AND_VERSION);
+    mp_hal_stdout_tx_str("; " MICROPY_BANNER_MACHINE);
+    mp_hal_stdout_tx_str("\nUse Ctrl-D to exit, Ctrl-E for paste mode\n");
+
     setvbuf(stdin,  NULL, _IONBF, 0);
     setvbuf(stdout, NULL, _IONBF, 0);
 
-    int stack_dummy;
+    // int stack_dummy;
```
Comment (jaenrig-ifx) [inline ports/psoc6/main.c:120]: Please, remove all commented block which are not needed. Or add some comment about what is the plan, when to uncomment, todo, etc.

File: ports/psoc6/tests/vfs_fat.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
+import os
```
Comment (jaenrig-ifx) [inline ports/psoc6/tests/vfs_fat.py:1]: Wouldn´t these test belong to test/psoc6 ? or even there is some generic file system tests that we can run?
Any changed required to be added to the automated check?

**Review by jaenrig-ifx — CHANGES_REQUESTED** [review]
Very nice @bargfred ! The only thing is the commented code. Having commented code without any explanation we have to avoid. 
And the test scripts, which we should be able to move them to the tests folder, and we include them in our ci/cd run.

File: ports/psoc6/boards/manifest.py (hunk 1 above)
Comment (bargfred) [reply ports/psoc6/boards/manifest.py:2]: ok

File: ports/psoc6/main.c (hunk 2 above)
Comment (bargfred) [reply ports/psoc6/main.c:120]: ok

File: ports/psoc6/modules/gc/modgc.c [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+
+#if MICROPY_ENABLE_GC
+
+void gc_collect(void) {
+    printf("gc_collect\n");
+
+    // WARNING: This gc_collect implementation doesn't try to get root
+    // pointers from CPU registers, and thus may function incorrectly.
+//    void *dummy;
```
[the remark this answers is not in the document]
Comment (bargfred) [reply ports/psoc6/modules/gc/modgc.c:16]: ok

File: ports/psoc6/modules/os/moduos.c [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+STATIC mp_obj_t mp_uos_getenv(mp_obj_t var_in) {
+    const char *s = getenv(mp_obj_str_get_str(var_in));
+    if (s == NULL) {
+        return mp_const_none;
+    }
+    return mp_obj_new_str(s, strlen(s));
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_1(mp_uos_getenv_obj, mp_uos_getenv);
+
+// STATIC mp_obj_t mp_uos_putenv(mp_obj_t key_in, mp_obj_t value_in) {
```
[the remark this answers is not in the document]
Comment (bargfred) [reply ports/psoc6/modules/os/moduos.c:44]: ok

File: ports/psoc6/modules/psoc6/psoc6_flash.c [hunk 6]
```diff
@@ -0,0 +1,308 @@
...
+
+STATIC psoc6_flash_obj_t psoc6_flash_obj = {
+    .base = { &psoc6_flash_type },
+    .flash_base = FLASH_BASE,
+    .flash_size = FLASH_SIZE,
+};
+
+cyhal_flash_t cyhal_flash_obj;
+
+// // Tag the flash drive in the binary as readable/writable (but not reformatable)
```
[the remark this answers is not in the document]
Comment (bargfred) [reply ports/psoc6/modules/psoc6/psoc6_flash.c:79]: ok

File: ports/psoc6/tests/vfs_fat.py (hunk 3 above)
Comment (bargfred) [reply ports/psoc6/tests/vfs_fat.py:1]: These tests were thought to be made examples for the user, but are not yet finalized. Then they could also be used for run_tests.py. 
So, shall I remove until a final version is available ?

Comment (jaenrig-ifx) [reply ports/psoc6/tests/vfs_fat.py:1]: Mmm I would not remove them...  
I think the primary user, unless pointed to the repo would not look for examples here. But mpy developers or maintainers could take the tests as examples... 
At the same time, a primary user example does not necessarily needs to cover all the module functionality. And in a test, we might end having also some verification of not sunny side paths, forcing failures and so on. So some parts might not be useful in terms of explaining how to use the module to the user... But still, I would consider for now the tests as a big example for the complete (or as complete as possible) functionality of the module.

Comment (bargfred) [reply ports/psoc6/tests/vfs_fat.py:1]: Agreed, but now what next ?

Comment (jaenrig-ifx) [reply ports/psoc6/tests/vfs_fat.py:1]: If they are ready to be run under ci/cd, I would add them to the test folder. If not we can just keep them where they are and do that in an upcoming ticket ?

**Review by jaenrig-ifx — APPROVED** [review]

File: ports/psoc6/main.c [hunk 7]
```diff
@@ -22,38 +22,59 @@
...
-        mp_call_function_0(module_fun);
-        nlr_pop();
-    } else {
-        // uncaught exception
-        mp_obj_print_exception(&mp_plat_print, (mp_obj_t)nlr.ret_val);
-    }
-}
-
-#endif
+cyhal_rtc_t psoc6_rtc;
```
Comment (boramonideep) [inline ports/psoc6/main.c:29]: Just an abstract design level question/suggestion: 

Should we expose `cy_/hal/pdl` related macros and types to the main program? I and Nikhita also discussed about this in the end decided to encapsulate such `cy/hal_pdl` in a way defined here: https://github.com/jaenrig-ifx/micropython/blob/82a146a2f02be2ba67c5698d4a720dbc8d9d4959/ports/psoc6/drivers/psoc6_gpio.h#L18

File: ports/psoc6/main.c [hunk 8]
```diff
@@ -67,6 +88,7 @@ int main(int argc, char **argv) {
 
     #endif /* #if defined (CY_DEVICE_SECURE) */
 
+
     /* Initialize the device and board peripherals */
     result = cybsp_init();
```
Comment (boramonideep) [inline ports/psoc6/main.c:93]: Similar to https://github.com/jaenrig-ifx/micropython/pull/17/files#r1049659473 .

File: ports/psoc6/modules/psoc6/psoc6_fatfs.c [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+extern cyhal_rtc_t psoc6_rtc;
+
+
+MP_WEAK DWORD get_fattime(void) {
+    struct tm current_date_time = {0};
+    cy_rslt_t result = cyhal_rtc_read(&psoc6_rtc, &current_date_time);
+
+    if (CY_RSLT_SUCCESS != result) {
+        // mp_raise_NotImplementedError("cyhal_rtc_read failed !");
+        printf("cyhal_rtc_read failed !");
```
Comment (boramonideep) [inline ports/psoc6/modules/psoc6/psoc6_fatfs.c:46]: As an alternative to a simple printf MPY provides also a function to raise custom exceptions like here: 
https://github.com/jaenrig-ifx/micropython/blob/82a146a2f02be2ba67c5698d4a720dbc8d9d4959/ports/psoc6/modules/machine/machine_pin.c#L196

Comment (bargfred) [reply ports/psoc6/modules/psoc6/psoc6_fatfs.c:46]: I changed it to what it have should been : mp_raise_NotImplementedError(MP_ERROR_TEXT(...))

**Comment by boramonideep** [comment]
Additional comment: Since we already have a preliminary debug logger now. We could use that instead of commenting out the print statements. The logger in its current state is a little expensive (when it is inactive) but I have some ideas to improve on that. I could not find any particular instance in this PR but just a general suggestion.

The changes will be tremendously helpful for the future implementations. Thank you @bargfred

**Comment by bargfred** [comment]
Will do that now that the logger is available, but do not want to do that for this pull request.

**Review by boramonideep — APPROVED** [review]

**Review by IFX-Anusha — APPROVED** [review]

**Review by NikhitaR-IFX — COMMENTED** [review]
I am parallelly pulling your changes to use them (time module). I upvote for having "examples/machine" and including basic examples for modules we develop.

**Review by NikhitaR-IFX — APPROVED** [review]

**Pull request merged**